grpc-client-server/README.md

```diff
@@ -0,0 +1,188 @@
+# GRPC Client and Server Communication 
+
+The goals of this project are: 
+ * to use gRPC requests, and understand how they differ from REST
+ * to use deadlines with gRPCs
+ * to build a simple prober
+ * optionally, gain experience with instrumenting an application and running Prometheus
+
+## What is gRPC?
+
+RPCs aren’t the same as HTTP requests. HTTP is a text-based protocol, oriented towards resources or entities. RPCs are oriented towards calling a specific function. RPCs are highly structured; normally the request and response are encoded in efficient binary formats which are not human-readable. HTTP APIs are usually limited to CRUD (Create, Read, Update, Delete) operations and RPCs can perform any kind of operation. If you want to efficiently integrate two systems that you control, RPCs are a good choice. If you want to provide an API for use by developers outside of your organisation then HTTP APIs are generally a better choice, because they are simpler to develop against and all programming languages provide good HTTP support. gRPC, which we will use in this exercise, is a RPC implementation that is fairly common in industry. 
+
+Read the [gRPC Introduction](https://grpc.io/docs/what-is-grpc/introduction/) and 
+[gRPC Core Concepts](https://grpc.io/docs/what-is-grpc/core-concepts/) for an overview of gRPC.
+todo stuff from primer
+
+## Get started with a HelloWorld gRPC example
+
+Start by building a simple gRPC ‘hello world’ server and cli client in golang. See [gRPC Quickstart](https://grpc.io/docs/languages/go/quickstart/) for detailed instructions.
+
+## Change HelloWorld to Prober
+
+Once you have the 'hello world' example from the gRPC Quickstart tutorial working, we will move on to implementing a prober.
+
+Rename `greeter_client` and `greeter_server` to `prober_client` and `prober_server`.   
+Change the `helloworld` subdirectory to be called `prober`,` and helloworld.proto` to `prober.proto`.
+Delete the genereted .go files in the same directory as `prober.proto`.
+
+Now update the the prober.proto file so that the main content section looks like this:
+
+```
+package prober;
+
+// The greeting service definition.
+service Prober {
+  // Sends a greeting
+  rpc DoProbes (ProbeRequest) returns (ProbeReply) {}
+}
+
+// The request message containing the user's name.
+message ProbeRequest {
+  string name = 1;
+}
+
+// The response message containing the greetings
+message ProbeReply {
+  string message = 1;
+}
+```
+
+Regenerate the generated protobuf code: 
+```
+protoc --go_out=. --go_opt=paths=source_relative \
+    --go-grpc_out=. --go-grpc_opt=paths=source_relative prober/prober.proto
+```
+
+Observe the new generated files:
+```
+ls prober
+prober.pb.go		prober.proto		prober_grpc.pb.go
+```
+
+Read through `prober_grpc.pb.go` - this is the interface you will use in your code. This is how gRPC works: a `proto` format 
```

Comment:
Stylistic but `you` and `your` → to `we` and `our` please, unless giving direct instructions (e.g. `You need to set up account`, `You should not...`). I can see you've used "we" elsewhere.

Note to self: need to add this to `CONTRIBUTING.md`.